### PR TITLE
Improve parsing large integers

### LINKED ISSUE #980: Extremely large integers have slow parsing performance in toolchain
Large integers (10k+ digits) have sufficiently poor parsing performance that we're seeing it cause timeouts for the toolchain fuzzers. 

The slow code seems to be in getAsInteger, called from numeric_literal.cpp ([permalink](https://github.com/carbon-language/carbon-lang/blob/e4caf7d604b288796d7e1b92b0d55535846f43c8/toolchain/lex/numeric_literal.cpp#L215))

From some digging, I think the issue is that APInt has visibly slow performance when it's dealing with these large integers.

Examples can be found at https://github.com/carbon-language/carbon-lang/pull/979/files

One solution I'm considering is just having fuzzers examine input for these large numbers, and discard them. e.g., exit quickly if fuzz input contains 1k digits. But I may take a quick look to see if I can improve parsing performance first.

### PR DESCRIPTION
**Pull request by jonmeow**
Though I started this thinking about performance of parse of large decimal integers, I extended it to generally improve performance of integer values (TBH I hadn't expected such a difference for binary/hex, but I'll take it).

Note I think tests change because I'm making subtle changes to bit widths. The changes themselves appear harmless to me, but happy to make changes if it'd help.

Bumping up the number of digits by 10x because it's not really a performance issue anymore (eh, maybe somebody will want to specify a 256-byte value in binary). But, at a certain point it still seems like a mistake if somebody has that many digits in a row.

Fixes #980

Highlighting benchmark differences:
```diff
- BM_ComputeValue_IntDecimalN/1           37.1 ns         37.1 ns     18887116
+ BM_ComputeValue_IntDecimalN/1           21.9 ns         21.9 ns     31902433
- BM_ComputeValue_IntDecimalN/10000 1251228680 ns   1250457559 ns            1
+ BM_ComputeValue_IntDecimalN/10000     458818 ns       458626 ns         1523
- BM_ComputeValue_IntBinaryN/1            29.0 ns         29.0 ns     24058533
+ BM_ComputeValue_IntBinaryN/1            22.2 ns         22.1 ns     31566949
- BM_ComputeValue_IntBinaryN/10000     1390557 ns      1389782 ns          506
+ BM_ComputeValue_IntBinaryN/10000       16402 ns        16396 ns        42744
- BM_ComputeValue_IntHexN/1               34.0 ns         34.0 ns     20562432
+ BM_ComputeValue_IntHexN/1               22.4 ns         22.4 ns     31238055
- BM_ComputeValue_IntHexN/10000        5387942 ns      5385262 ns          130
+ BM_ComputeValue_IntHexN/10000          39249 ns        39233 ns        17859
```

Benchmark before:
```
----------------------------------------------------------------------------
Benchmark                                  Time             CPU   Iterations
----------------------------------------------------------------------------
BM_Lex_Float                            10.6 ns         10.6 ns     66138191
BM_Lex_Int                              15.5 ns         15.4 ns     45149703
BM_Lex_IntDecimalN/1                    3.11 ns         3.11 ns    225524908
BM_Lex_IntDecimalN/10                   11.8 ns         11.8 ns     56719805
BM_Lex_IntDecimalN/100                   102 ns          102 ns      6867468
BM_Lex_IntDecimalN/1000                  943 ns          942 ns       745313
BM_Lex_IntDecimalN/10000                9465 ns         9461 ns        73970
BM_ComputeValue_Float                   61.6 ns         61.6 ns     11377463
BM_ComputeValue_Int                      106 ns          106 ns      6587381
BM_ComputeValue_IntDecimalN/1           37.1 ns         37.1 ns     18887116
BM_ComputeValue_IntDecimalN/10          87.7 ns         87.7 ns      7960837
BM_ComputeValue_IntDecimalN/100         7963 ns         7956 ns        88858
BM_ComputeValue_IntDecimalN/1000     1212577 ns      1211906 ns          578
BM_ComputeValue_IntDecimalN/10000 1251228680 ns   1250457559 ns            1
BM_ComputeValue_IntBinaryN/1            29.0 ns         29.0 ns     24058533
BM_ComputeValue_IntBinaryN/10           69.4 ns         69.4 ns     10108642
BM_ComputeValue_IntBinaryN/100           963 ns          962 ns       726982
BM_ComputeValue_IntBinaryN/1000        21562 ns        21551 ns        32506
BM_ComputeValue_IntBinaryN/10000     1390557 ns      1389782 ns          506
BM_ComputeValue_IntHexN/1               34.0 ns         34.0 ns     20562432
BM_ComputeValue_IntHexN/10              70.4 ns         70.4 ns      9953165
BM_ComputeValue_IntHexN/100             1474 ns         1473 ns       472776
BM_ComputeValue_IntHexN/1000           61818 ns        61762 ns        11363
BM_ComputeValue_IntHexN/10000        5387942 ns      5385262 ns          130
```

Benchmark after:
```
----------------------------------------------------------------------------
Benchmark                                  Time             CPU   Iterations
----------------------------------------------------------------------------
BM_Lex_Float                            10.9 ns         10.9 ns     63993114
BM_Lex_Int                              15.1 ns         15.1 ns     46869766
BM_Lex_IntDecimalN/1                    3.16 ns         3.16 ns    220923300
BM_Lex_IntDecimalN/10                   12.2 ns         12.2 ns     57731654
BM_Lex_IntDecimalN/100                   102 ns          102 ns      6875516
BM_Lex_IntDecimalN/1000                  942 ns          942 ns       742359
BM_Lex_IntDecimalN/10000                9353 ns         9350 ns        75096
BM_ComputeValue_Float                   44.9 ns         44.9 ns     15619691
BM_ComputeValue_Int                     48.9 ns         48.9 ns     14361507
BM_ComputeValue_IntDecimalN/1           21.9 ns         21.9 ns     31902433
BM_ComputeValue_IntDecimalN/10          30.3 ns         30.3 ns     23134117
BM_ComputeValue_IntDecimalN/100          224 ns          223 ns      3092567
BM_ComputeValue_IntDecimalN/1000        5834 ns         5830 ns       117469
BM_ComputeValue_IntDecimalN/10000     458818 ns       458626 ns         1523
BM_ComputeValue_IntBinaryN/1            22.2 ns         22.1 ns     31566949
BM_ComputeValue_IntBinaryN/10           32.9 ns         32.9 ns     21306927
BM_ComputeValue_IntBinaryN/100           198 ns          198 ns      3545277
BM_ComputeValue_IntBinaryN/1000         1671 ns         1669 ns       419656
BM_ComputeValue_IntBinaryN/10000       16402 ns        16396 ns        42744
BM_ComputeValue_IntHexN/1               22.4 ns         22.4 ns     31238055
BM_ComputeValue_IntHexN/10              47.8 ns         47.7 ns     14694407
BM_ComputeValue_IntHexN/100              436 ns          436 ns      1609794
BM_ComputeValue_IntHexN/1000            3966 ns         3962 ns       177109
BM_ComputeValue_IntHexN/10000          39249 ns        39233 ns        17859
```

Assisted-by: Google Antigravity with Gemini

### CODE REVIEW

File: toolchain/lex/numeric_literal.cpp [hunk 1]
```diff
@@ -176,6 +177,92 @@ auto NumericLiteral::Parser::Check() -> bool {
...
+  return chunk_val;
+}
+
+// Parsing decimals is complex because they're not a power of 2. We process it
+// 19 digits at a time because that's the most that fit into uint64_t, which
+// is APInt's internal unit for storage; chunking this way minimizes
+// cross-unit arithmetic.
+static auto ParseDecimal(llvm::StringRef digits) -> llvm::APInt {
+  // APInt performance scales based on the number of bits, so be precise.
+  static const double bits_per_digit = std::log2(10);
```
Comment (danakj) [inline toolchain/lex/numeric_literal.cpp:235]: TODO: constexpr in C++26?

Or should we just write out the constant result by hand?

File: toolchain/lex/numeric_literal_test.cpp
Comment (danakj) [inline toolchain/lex/numeric_literal_test.cpp:116]: Are we missing test coverage where the non-first chunks have non-zero values?

**Review by danakj — APPROVED** [review]
LGTM

File: toolchain/lex/numeric_literal.cpp (hunk 1 above)
Comment (jonmeow) [reply toolchain/lex/numeric_literal.cpp:235]: > TODO: constexpr in C++26?

Added, though I admit I hope this is moot. On the one hand, Carbon's still using C++20... That was adopted [in 2024](https://github.com/carbon-language/carbon-lang/pull/3660), so if C++26 is also a 4-year delay, I hope Carbon is self-hosting before then. :)

> Or should we just write out the constant result by hand?

For me, the calculation is more recognizable than the value (also avoids precision), and I don't think this is a performance issue as a static.

File: toolchain/lex/numeric_literal_test.cpp
Comment (jonmeow) [reply toolchain/lex/numeric_literal_test.cpp:116]: Added a test in numeric_literals.carbon, which also shows the before/after (i.e. not just that it parsed, but what it parsed to)

**Review by danakj — APPROVED** [review]
LGTM

**Comment by niekbouman** [comment]
Hello @jonmeow, 

We had a discussion in https://github.com/carbon-language/carbon-lang/issues/980 a long time ago.
Today, I was experimenting with compile-time parsing of integer literals in C++ using Gemini, 
and then I had to think back to our conversation, at noticed that you in fact worked on this two days ago.

The code that resulted from this experiment I did is perhaps valuable to you as well.
It also groups 19 decimal digits together, and then does a tree-fold over the words. 
I actually use the _BitInt type (which currently only works in Clang as a non-standard extension, but this might be standardized in c++29)
The tree fold is implemented with recursion, and the bitwidth of the _BitInt type is chosen as a function of the depth in the tree. 

If you're interested, have a look at https://godbolt.org/z/9rhjao498
best,
Niek

**Comment by jonmeow** [comment]
@niekbouman I'd experimented with a tree-like multiply with APInt. However, `APInt * APInt` multiplies all bytes in both values, and both operands must be equal bit width; in contrast, `APInt * u64` does multiplication of different bit widths, and so is able to do fewer multiplies. I think this is where it led to generally worse performance.

I will note, I didn't investigate whether copying values in order to have varying bit widths (and thus cheaper multiplies) would have a significant difference. The performance even for short bit widths (where this wouldn't have much impact) looked worse. I leaned more towards simplicity and that the current performance is good enough.

For `_BitInt`, note we still need `APInt` for Clang/LLVM so any use would still need to be copied over. Also, it's not clear to me that `_BitInt` would avoid the problems that `APInt` encounters... it may do better for being compile-time sized, but also that means more complexity in the implementation because it needs to efficiently parse arbitrarily-sized values.

Anyways, it'd probably take more investigation into relative performance. If you're interested, the benchmarks are there for comparison now. :)
